lib/logdna/client.rb

```diff
@@ -1,164 +1,130 @@
+      @flush_scheduled = false
+      @exception_flag = false
 
-      begin
-          msg = msg.encode("UTF-8")
-      rescue Encoding::UndefinedConversionError => e
-        # NOTE: should this be raised or handled silently?
-        # raise e
-      end
-      msg
+      @request = request
+      @retry_timeout = opts[:retry_timeout] ? opts[:retry_timeout] : Resources::RETRY_TIMEOUT
     end
 
-    def message_hash(msg, opts={})
-      obj = {
+    def process_message(msg, opts = {})
+      processed_message = {
         line: msg,
         app: opts[:app],
         level: opts[:level],
         env: opts[:env],
         meta: opts[:meta],
         timestamp: Time.now.to_i,
       }
-      obj.delete(:meta) if obj[:meta].nil?
-      obj
-    end
-
-    def create_flush_task
-      return @task unless @task.nil? or !@task.running?
-
-      t = Concurrent::TimerTask.new(execution_interval: @actual_flush_interval, timeout_interval: Resources::TIMER_OUT) do |task|
-        if @messages.any?
-          # keep running if there are queued messages, but don't flush
-          # because the buffer is being flushed due to being over the limit
-          unless @buffer_over_limit
-            flush()
-          end
-        else
-          # no messages means we can kill the task
-          task.kill
-        end
-      end
-      t.execute
-    end
-
-    def check_side_buffer
-      return if @side_buffer.size == 0
-
-      @buffer.write(@side_buffer.string)
-      @side_buffer.truncate(0)
-      queued_side_messages = @side_messages
-      @side_messages = []
-      queued_side_messages.each { |message_hash_obj| @messages.push(message_hash_obj) }
+      processed_message.delete(:meta) if processed_message[:meta].nil?
+      processed_message
     end
 
-
-    # this should always be running synchronously within this thread
-    def buffer(msg, opts)
-      buffer_size = write_to_buffer(msg, opts)
-      unless buffer_size.nil?
-        process_buffer(buffer_size)
-      end
+    def schedule_flush
+      @flush_scheduled = true
+      start_timer = lambda {
+        sleep(@exception_flag ? @retry_timeout : @flush_interval)
+        flush
+      }
+      thread = Thread.new { start_timer }
+      thread.join
     end
 
     def write_to_buffer(msg, opts)
-      return if msg.nil?
-      msg = encode_message(msg)
+      if @lock.try_lock
+        @buffer.concat(@side_messages) unless @side_messages.empty?
+        processed_message = process_message(msg, opts)
+        new_message_size = processed_message.to_s.bytesize
 
-      if @lock.locked?
-        @side_buffer.write(msg)
-        @side_messages.push(message_hash(msg, opts))
-        return
-      end
+        @buffer_byte_size += new_message_size
+        @buffer.push(processed_message)
 
-      check_side_buffer
-      buffer_size = @buffer.write(msg)
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
+        @lock.unlock if @lock.locked?
```

Comment:
OK, having read through this all, I now see why you did the unlock here and to me that's an indication that the lock logic is not right - it really should be locked/unlocked in roughly the same block - there's a somewhat crazy coordination game going on right now that makes the logic unclear, and I think not quite right.